File: src/algorithms/dynamic-programming/integer-break-memoization/sources/integer-break-memoization.ts
// Integer Break memoization — top-down recursion to maximize product of parts

export function integerBreakMemoization(
  targetNumber: number,
  memo: Map<number, number> = new Map(),
): number {
  // @step:initialize
  if (targetNumber === 1) return 1; // @step:initialize

  if (memo.has(targetNumber)) return memo.get(targetNumber)!; // @step:read-cache
  // @step:push-call
  let maxProduct = 0; // @step:compute-cell
  for (let partSize = 1; partSize < targetNumber; partSize++) {
    // @step:compute-cell
    const remainder = targetNumber - partSize; // @step:compute-cell
    const splitProduct = partSize * remainder; // @step:compute-cell
    const recurseProduct = partSize * integerBreakMemoization(remainder, memo); // @step:compute-cell
    maxProduct = Math.max(maxProduct, splitProduct, recurseProduct); // @step:compute-cell
  }
  memo.set(targetNumber, maxProduct); // @step:compute-cell
  return maxProduct; // @step:pop-call
}

File: src/algorithms/dynamic-programming/optimization/integer-break-memoization/sources/IntegerBreakMemoization.cpp
// Integer Break memoization — top-down recursion to maximize product of parts

#include <iostream>
#include <unordered_map>
#include <algorithm>

int integerBreakMemoization(int targetNumber, std::unordered_map<int, int>& memo) {
    // @step:initialize
    if (targetNumber == 1) return 1; // @step:initialize
    auto it = memo.find(targetNumber);
    if (it != memo.end()) return it->second; // @step:read-cache
    // @step:push-call
    int maxProduct = 0; // @step:compute-cell
    for (int partSize = 1; partSize < targetNumber; partSize++) {
        // @step:compute-cell
        int remainder = targetNumber - partSize; // @step:compute-cell
        int splitProduct = partSize * remainder; // @step:compute-cell
        int recurseProduct = partSize * integerBreakMemoization(remainder, memo); // @step:compute-cell
        maxProduct = std::max({maxProduct, splitProduct, recurseProduct}); // @step:compute-cell
    }
    memo[targetNumber] = maxProduct; // @step:compute-cell
    return maxProduct; // @step:pop-call
}

int main() {
    std::unordered_map<int, int> memo;
    int targetNumber = 10;
    int result = integerBreakMemoization(targetNumber, memo);
    std::cout << "Integer break(" << targetNumber << "): " << result << std::endl;
    return 0;
}
